### ISSUE TITLE
Switching DBs to support a multi-tenant environment

### ISSUE THREAD

**Issue by jfrumar** [balderdashy/sails-mongo#130]
We have had success with Sails up to the point where we need to switch DBs based on a requesting tenant's ID. Sails doesn't seem to support multi-tenanting operations easily.

As a proof of concept, I wrote a custom mongo adapter (ref: https://github.com/jfrumar/sails-mongo-multi) that would maintain multiple DB connections in memory and switch them (through the adapter's exposed 'switchDB' function), but it doesn't support concurrent connections (i.e. a second request can switch the DB from underneath the first request while it's still in flight which is very dangeous for a multi-tenanted system).

I believe the only way to support this will be to pass the express js 'req' object all the way down to the adapter, since the tenant ID is contained within (Sails globals cannot be relied upon because they are shared between all the concurrent requests).

Is there a recommended approach for switching DBs on the fly based on each request, or is this something that's far off the roadmap of Sails and the mongo adapter?

**Comment by Glavin001** [comment]
## Update on the above proposal

I think that this could be implemented at a Waterline level.

The `sails-mongo` adapter has a `registerConnection` method that sets up the Database connection asyncronously ( see https://github.com/balderdashy/sails-mongo/blob/master/lib/adapter.js#L70-L113 ).

This `registerConnection` method is called by Waterline at
https://github.com/balderdashy/waterline/blob/63effdb480ae9bd61cd78fd534e1e513294bd98d/lib/waterline.js#L174-L208

I searched through the Waterline source code and found https://github.com/balderdashy/waterline/blob/master/lib/waterline/adapter/dql.js which I hope to be the bridge between Sails and Waterline. I hope that wrapping this methods in an asynchronous function that checks if `isMultiTenant` is enabled and if so then resolves the new tenant's connection before continuing to call the core method (create, find, etc).

/cc @eubruno

**Comment by Glavin001** [comment]
I have submitted a Pull Request: https://github.com/balderdashy/waterline/pull/787

**Comment by kheide** [comment]
By looking at these last suggestions it seems that multi-tenancy is tried solved by routing requests from sails to different db's. How would this be scalable in the long term? 

Would a solution with multi-tenancy in waterline be preferable? And then run a single-db setup. How would this be security-wise?

**Comment by Glavin001** [comment]
@kheide I agree and had implemented a solution in Waterline: https://github.com/balderdashy/waterline/pull/787
It will not be merged, even though it meets all of my needs. After discussion, there were other objectives that needed to be addressed, so looks like it will be a while before Sails / Waterline has multi-tenancy support.

**Comment by kheide** [comment]
I've read all of your discussions. I am insecure about using your approach in our project since it's not in core and thus not supported. And it would suck to be stranded running into issues when we are live with customers. 

What I am interested in doing is grab the sub-domain from the request (which would identify the tenant) and route that to a separate DB. Basically dynamically changing the 'database' field in connections.js 

Would something like this be at all feasible? Is there an easier solution to my simple approach?

**Comment by Glavin001** [comment]
My solution does exactly what you described, with dynamic databases and more based off of the incoming request. 




I understand and agree. I do not recommend using my solution in production, until it is merged upstream. 




—

**Comment by kheide** [comment]
I really do hope either you come back to this one or the devs pick it up. I would hate to have to migrate over to rails just for the multi-tenancy when everything else about Sails rocks!

**Comment by sailsbot** [comment]
Thanks for posting, @jfrumar.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by smrutimandal** [comment]
Hi guys,

I am in the middle of writing a multi-tenant applications have the following usecase(database engine is pgsql 9.4):

 1. Sails lifts with a seed database.
 2. While the application is online, new databases may be added and some removed. Hence, I cannot pre-configure my connections file at the time of startup.
 3. I can figure out from my policy which database I need to hit to get the data.(All databases have the same exact tables, they differ only in data)

I have read up a bit on this in StackOverflow and GitHub(all great discussions in this area) and haven't been able to find anything that WORKS.

Several proposed Solutions
 - Multiple connections in the file
	Sails doesn't won't know before hand how many databases it needs to interact with at the time of coming up.
 - change connections file and restart app
	This is not a solution that works for me as the DBs are added and removed all the time. I would like to run multiple instances of a single app in dockers and it is not feasible to edit the connections.js file and restart the app/orm in a batch of dockers at a time.
 - Switch to native('pg') adapter & craft the connections manually.
	I like the way Waterline smoothens things up in Sails code and going the native way will be one of my last choices.

I found an indication about on upcoming feature similar to this
http://stackoverflow.com/questions/29454176/sails-js-i-want-to-add-db-connection-dynamically-after-sails-lift/34845674#34845674

I can also see version 1.0 roadmap has features(e.g. Built-in Support for Dynamic Database Connections - sails.hooks.orm.datastore()) in these areas taken into consideration. It apparently has been implemented at the driver level as of October. If someone can guide me in taking in these changes into sails/waterline I can ok with living on the bleeding edge.

Kindly let me know if there is a way to let the controller code decide which database to execute the oprations in.

Stack
sails : 0.12.11
sails-postgresql: 0.11.4

I would also like thank this community who can enable newbies like to me to write 10s of thousands of line codes in a matter of months. Keep up the good work guys!
